File: Linked list/add_two_numbers.py
"""You are given two non-empty linked lists representing two non-negative 
integers. The digits are stored in reverse order and each of their nodes 
contain a single digit. Add the two numbers and return it as a linked list.
You may assume the two numbers do not contain any leading zero, except 
the number 0 itself."""

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
        
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        cur1 = l1
        cur2 = l2
        head = ListNode(0)
        temp = head
        dep = 0
        while cur1 is not None and cur2 is not None:
            sm = cur1.val + cur2.val + dep
            if sm < 10:
                temp.next = ListNode(sm)
                temp = temp.next
                dep = 0
            else:
                fr = sm%10
                temp.next = ListNode(fr)
                temp = temp.next
                dep = 1
            cur1 = cur1.next
            cur2 = cur2.next
        while cur1 is not None:
            sm = dep + cur1.val
            if sm > 9:
                fr  = sm%10
                temp.next = ListNode(fr)
                temp = temp.next
                dep =1
            else:
                temp.next = ListNode(sm)
                temp = temp.next
                dep = 0
            cur1 = cur1.next
        
        while cur2 is not None:
            sm = dep + cur2.val
            if sm > 9:
                fr  = sm%10
                temp.next = ListNode(fr)
                temp = temp.next
                dep =1
            else:
                temp.next = ListNode(sm)
                temp = temp.next
                dep = 0
            cur2 = cur2.next
        if dep == 1:
            temp.next = ListNode(dep)
        return head.next
            

File: Linked list/AddTwoNumbers.cpp
//Problem Link:- https://leetcode.com/problems/add-two-numbers/submissions/

#include<bits/stdc++.h>
using namespace std;


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* first = l1;
        ListNode* sec = l2;
        ListNode* ans = new ListNode(-1);
        ListNode* res = ans;
        int carry = 0;
        int sm = 0;
        while (first != NULL && sec != NULL){
            sm = first->val + sec->val + carry;
            carry = sm / 10;
            int dg = sm % 10;
            ans->next = new ListNode(dg);
            ans = ans->next;
            first = first->next;
            sec = sec->next;
        }
        while (first != NULL){
            sm = carry + first->val;
            carry = sm / 10;
            int dg = sm % 10;
            ans->next = new ListNode(dg);
            ans = ans->next;
            first = first->next;
        }
        while (sec != NULL){
            sm = carry + sec->val;
            carry = sm / 10;
            int dg = sm  % 10;
            ans->next = new ListNode(dg);
            ans = ans->next;
            sec = sec->next;
        }
        if (carry > 0){
            ans->next = new ListNode(carry);
        }
        return res->next;
    }
};